src/systems/detachable_joint/DetachableJoint.cc

```diff
@@ -234,8 +234,75 @@ void DetachableJoint::PreUpdate(
     }
     else
     {
-      modelEntity = _ecm.EntityByComponents(
-          components::Model(), components::Name(this->childModelName));
+      auto entitiesMatchingName = entitiesFromScopedName(
+        this->childModelName, _ecm);
+
+      // TODO(arjoc): There is probably a more efficient way of combining entitiesFromScopedName
+      // With filtering.
+      // Filter for entities with only models
+      std::vector<Entity> candidateEntities;
+      for (const auto entity: entitiesMatchingName)
+      {
+        if (_ecm.EntityHasComponentType(entity, components::Model::typeId))
+        {
+          candidateEntities.push_back(entity);
+        }
+        else
+        {
+          ignwarn << "Found entity " << scopedName(entity, _ecm)
+            << " but it is not a model." << std::endl;
+        }
+      }
+
+      if (candidateEntities.size() == 1)
+      {
+        // If there is one entity select that entity itself
+        modelEntity = *candidateEntities.begin();
+      }
+      else
+      {
+        std::string selectedModelName;
+        auto parentEntityScopedPath = scopedName(this->model.Entity(), _ecm);
+        // If there is more than one entity with the same name, look for the
+        // entity with a parent component.
+        for (auto entity : candidateEntities)
+        {
+          auto childEntityScope = scopedName(entity, _ecm);
+          if (childEntityScope.size() < parentEntityScopedPath.size())
+          {
+            continue;
+          }
+          if (childEntityScope.find(parentEntityScopedPath) != 0)
```

Comment:
```suggestion
          if (childEntityScope.rfind(parentEntityScopedPath, 0) != 0)
```
This version is more efficient as it restricts the search to a prefix in childEntityScope.

Reply:
[1a7083d](https://github.com/gazebosim/gz-sim/pull/1097/commits/1a7083dfeb05bd50e6358bbbd3377f6a13e0bb4e)